### ISSUE TITLE
Help needed to explain the orchestration reentrant behavior

### ISSUE THREAD

**Issue by t-bzhan** [Azure/durabletask#411]
We plan to use durable task framework, but feel it hard to understand the reentrant behavior of the orchestration. With below code snippets for a dummy orchestration/activity.

```
    public class DummyOrchestration : TaskOrchestration<string, string>
    {
        public override async Task<string> RunTask(OrchestrationContext context, string input)
        {
            Console.WriteLine($"{DateTimeOffset.Now}: DummyOrchestration {this.GetHashCode() } has received request.");
            Guid guid = await context.ScheduleTask<Guid>(typeof(DummyActivity), input);
            Console.WriteLine($"{DateTimeOffset.Now}: DummyOrchestration {this.GetHashCode() } has scheduled tasks with tracking id {guid}");

            for (int i = 0; i < 5; i++)
            {
                await context.CreateTimer(context.CurrentUtcDateTime.Add(TimeSpan.FromSeconds(3)), "timer1");
                Console.WriteLine($"{DateTimeOffset.Now}: DummyOrchestration {this.GetHashCode() } has slept for 3 seconds");
            }
            
            Console.WriteLine($"{DateTimeOffset.Now}: DummyOrchestration {this.GetHashCode() } has returned successful");

            return "successful";
        }
    }
```
```
    public sealed class DummyActivity : AsyncTaskActivity<string, Guid>
    {
        protected override async Task<Guid> ExecuteAsync(TaskContext context, string input)
        {
            await Task.Delay(new Random().Next(50, 100));
            Guid trackingId = Guid.NewGuid();
            Console.WriteLine($"{DateTimeOffset.Now}: Dummy {input} tracked with {trackingId}");

            return trackingId;
        }
    }
```
The output indicates clearly that the DummyOrchestration is initialized multiple times and partial of the remaining method of RunTask are not executed for all except the last instance, what confuses us most is how does it achieve this? Is it related to [SynchronousTaskScheduler](https://github.com/Azure/durabletask/blob/f9cc450539b5e37c97c19ae393d5bb1564fda7a8/src/DurableTask.Core/SynchronousTaskScheduler.cs) class?
```
6/27/2020 7:30:12 PM +08:00: DummyOrchestration 59884855 has received request.
6/27/2020 7:30:13 PM +08:00: Dummy entity1 tracked with 655fb40c-8c5f-4e52-adf0-ec7c09ffc37d
6/27/2020 7:30:14 PM +08:00: DummyOrchestration 48209832 has received request.
6/27/2020 7:30:14 PM +08:00: DummyOrchestration 48209832 has scheduled tasks with tracking id 655fb40c-8c5f-4e52-adf0-ec7c09ffc37d
6/27/2020 7:30:16 PM +08:00: DummyOrchestration 47129288 has received request.
6/27/2020 7:30:16 PM +08:00: DummyOrchestration 47129288 has scheduled tasks with tracking id 655fb40c-8c5f-4e52-adf0-ec7c09ffc37d
6/27/2020 7:30:16 PM +08:00: DummyOrchestration 47129288 has slept for 3 seconds
6/27/2020 7:30:19 PM +08:00: DummyOrchestration 62122949 has received request.
6/27/2020 7:30:19 PM +08:00: DummyOrchestration 62122949 has scheduled tasks with tracking id 655fb40c-8c5f-4e52-adf0-ec7c09ffc37d
6/27/2020 7:30:19 PM +08:00: DummyOrchestration 62122949 has slept for 3 seconds
6/27/2020 7:30:19 PM +08:00: DummyOrchestration 62122949 has slept for 3 seconds
6/27/2020 7:30:22 PM +08:00: DummyOrchestration 37919899 has received request.
6/27/2020 7:30:22 PM +08:00: DummyOrchestration 37919899 has scheduled tasks with tracking id 655fb40c-8c5f-4e52-adf0-ec7c09ffc37d
6/27/2020 7:30:22 PM +08:00: DummyOrchestration 37919899 has slept for 3 seconds
6/27/2020 7:30:22 PM +08:00: DummyOrchestration 37919899 has slept for 3 seconds
6/27/2020 7:30:22 PM +08:00: DummyOrchestration 37919899 has slept for 3 seconds
6/27/2020 7:30:25 PM +08:00: DummyOrchestration 63403007 has received request.
6/27/2020 7:30:25 PM +08:00: DummyOrchestration 63403007 has scheduled tasks with tracking id 655fb40c-8c5f-4e52-adf0-ec7c09ffc37d
6/27/2020 7:30:25 PM +08:00: DummyOrchestration 63403007 has slept for 3 seconds
6/27/2020 7:30:25 PM +08:00: DummyOrchestration 63403007 has slept for 3 seconds
6/27/2020 7:30:25 PM +08:00: DummyOrchestration 63403007 has slept for 3 seconds
6/27/2020 7:30:25 PM +08:00: DummyOrchestration 63403007 has slept for 3 seconds
6/27/2020 7:30:28 PM +08:00: DummyOrchestration 10508147 has received request.
6/27/2020 7:30:28 PM +08:00: DummyOrchestration 10508147 has scheduled tasks with tracking id 655fb40c-8c5f-4e52-adf0-ec7c09ffc37d
6/27/2020 7:30:28 PM +08:00: DummyOrchestration 10508147 has slept for 3 seconds
6/27/2020 7:30:28 PM +08:00: DummyOrchestration 10508147 has slept for 3 seconds
6/27/2020 7:30:28 PM +08:00: DummyOrchestration 10508147 has slept for 3 seconds
6/27/2020 7:30:28 PM +08:00: DummyOrchestration 10508147 has slept for 3 seconds
6/27/2020 7:30:28 PM +08:00: DummyOrchestration 10508147 has slept for 3 seconds
6/27/2020 7:30:28 PM +08:00: DummyOrchestration 10508147 has returned successful
```

I used to guess it is due to the instances are recycled, but with a memory dump, I found another interesting thing:  It seems that the DummyOrchestration instances are not GCed even I force the full GC with below code:
```
 GC.Collect(2);
 GC.WaitForPendingFinalizers();
 GC.Collect(2);
```
The object reference chain seems to have something to do with the reentrant behavior, but I am not able to figure it out.
![image](https://user-images.githubusercontent.com/61817681/85921419-d7a5a980-b8ae-11ea-8a31-eb2eb4d4df83.png)

I doubted these instances are not recycled because of the this.GetHashCode() API call in RunTask, but even I comments out all the diagnostic logs, it still has the same problem.

Could anyone familiar with the framework design to help us understand:

1. How does the durable task framework achieve the orchestration reentrant behavior?
2. Will those orchestration instances not recycled cause memory leak?

**Comment by cgillum** [comment]
When you talk about reentrancy, are you referring to the replay behavior of orchestrations? That is explained here: https://github.com/Azure/durabletask/wiki/Writing-Task-Orchestrations

Regarding your second question, I made the same observation, though I wasn’t able to conclude definitively whether those tasks never get cleaned up. But basically this is caused by the fact that the orchestration task for each replay starts but never completes, except for the last one. You can see that this is the case if you look at the code in TaskOrchestrationExecutor.cs.

**Comment by t-bzhan** [comment]
Thanks @cgillum for the link, it helps a lot. With more deep dive into the code and some local debugging, I find that the continuation of orchestration's RunTask never finishes might due to:

At first, TaskOrchestrationExecutor did not await the Execute method which calls RunTask underlyingly. In that case, the control is yielded at the first unfinished await. 
https://github.com/Azure/durabletask/blob/e213c2c75f72a6530cbb34affc80368fd24d4e9d/src/DurableTask.Core/TaskOrchestrationExecutor.cs#L135
https://github.com/Azure/durabletask/blob/e213c2c75f72a6530cbb34affc80368fd24d4e9d/src/DurableTask.Core/TaskOrchestration.cs#L96

While the ScheduleTask implementation in RunTask will wait for its internal TaskCompletionSource to become ready, but it will never get set except the last instance. That explains why the continuations are not called:
https://github.com/Azure/durabletask/blob/ef5bf0d686de898b37c696a6a712ca3153378496/src/DurableTask.Core/TaskOrchestrationContext.cs#L99-L102

As for unrecycled orchestration instances, it looks like it was only reproduced when I generated the dump with visual studio diagnostic tools in the debug model. I am not sure whether this is by design or not, but I did not notice the same problem when I ran the program in command line, sorry for the confusion.
